exmg-paper-token-input.ts

```diff
@@ -0,0 +1,546 @@
+
+  /**
+   * Set to true to auto-validate the input value.
+   */
+  @property({type: Boolean, attribute: 'auto-validate'})
+  public autoValidate: boolean = false;
+
+  @property({type: Boolean})
+  public autofocus: boolean = false;
+
+  /**
+   * Set to true to disable this input.
+   */
+  @property({type: Boolean})
+  public disabled: boolean = false;
+
+  /**
+   * The error message to display when the input is invalid.
+   */
+  @property({type: String, attribute: 'error-message'})
+  public errorMessage?: string;
+
+  /**
+   * alwaysFloatLabel
+   */
+  @property({type: Boolean})
+  public alwaysFloatLabel: boolean = false;
+
+  /**
+   * Maximum number of tokens allowed in value
+   */
+  @property({type: Number, attribute: 'max-tokens'})
+  public maxTokens?: number;
+
+  /**
+   * Set to true to mark the input as required. If you're using PaperInputBehavior to
+   * implement your own paper-input-like element, bind this to
+   * the `<input is="iron-input">`'s `required` property.
+   */
+  @property({type: Boolean})
+  public required: boolean = false;
+
+  /**
+   * This field will be bind to the actual input field
+   */
+  @property({type: String})
+  private inputValue: string = '';
+
+  @property({type: Boolean})
+  public invalid: boolean = false;
+
+  @property({type: Boolean})
+  public inputFocused: boolean = false;
+
+  @property({type: Boolean})
+  private opened: boolean = false;
+
+  @query('#listbox')
+  private listBoxNode?: PaperListboxElement;
+
+  @query('#inputValue')
+  private inputValueNode?: HTMLInputElement;
+
+  @query('#inputWidthHelper')
+  private inputWidthHelperNode?: HTMLElement;
+
+  @query('#menu')
+  private menuElement?: PaperMenuButton;
+
+  constructor() {
+    super();
+    this.onIronInputKeyDown = this.onIronInputKeyDown.bind(this);
+    this.onIronInputValueChanged = this.onIronInputValueChanged.bind(this);
+    this.onWindowClick = this.onWindowClick.bind(this);
+    this.onPaperListBoxItemSelect = this.onPaperListBoxItemSelect.bind(this);
+    this.onPaperListBoxItemDeselect = this.onPaperListBoxItemDeselect.bind(this);
+    this.onInputContainerTap = this.onInputContainerTap.bind(this);
+    this.onInputFocusChanged = this.onInputFocusChanged.bind(this);
+    this.onInputContainerButtonTap = this.onInputContainerButtonTap.bind(this);
+    this.onPaperMenuVisibilityChanged = this.onPaperMenuVisibilityChanged.bind(this);
+
+    this.resetInput = this.resetInput.bind(this);
+    this.computeAlwaysFloatLabel = this.computeAlwaysFloatLabel.bind(this);
+  }
+
+  private indexOfItem(item: HTMLElement): number {
+    return this.listBoxNode!.items ? (this.listBoxNode!.items || []).indexOf(item) : -1;
+  }
+
+  private indexOfValue(value: SelectedValue): number {
+    return this.listBoxNode!.items!
+        .map((item: HTMLElement) => this.getPaperItemValue(item))
+        .indexOf(value);
+  }
+
+  /** EVENT HANDLERS */
+
+  private previousClickWasInside: boolean = false;
```

Comment:
isn't better to keep properties on top of class ?

Reply:
done